common/raw_hashtable.h

```diff
@@ -0,0 +1,1206 @@
+                                  "can't have more probe steps than groups.";
+#endif
+  }
+
+  auto index() const -> ssize_t { return i; }
+};
+
+// TODO: Evaluate keeping this outlined to see if macro benchmarks observe the
+// same perf hit as micros.
+template <typename InputKeyT, typename InputValueT>
+template <typename LookupKeyT>
+auto ViewImpl<InputKeyT, InputValueT>::LookupEntry(LookupKeyT lookup_key) const
+    -> EntryT* {
+  // Prefetch with a "low" temporal locality as we're primarily expecting a
+  // brief use of the storage and then to return to application code.
+  __builtin_prefetch(this->storage_, /*read*/ 0, /*low-locality*/ 1);
+
+  ssize_t local_size = size();
+  CARBON_DCHECK(local_size > 0);
+
+  uint8_t* local_metadata = metadata();
+  HashCode hash = HashValue(lookup_key, ComputeSeed());
+  auto [hash_index, tag] = hash.ExtractIndexAndTag<7>();
+
+  EntryT* local_entries = entries();
+  ProbeSequence s(hash_index, local_size);
+  do {
+    ssize_t group_index = s.index();
+    MetadataGroup g = MetadataGroup::Load(local_metadata, group_index);
+    auto metadata_matched_range = g.Match(tag);
+    if (LLVM_LIKELY(metadata_matched_range)) {
+      EntryT* group_entries = &local_entries[group_index];
+      auto byte_it = metadata_matched_range.begin();
+      auto byte_end = metadata_matched_range.end();
+      do {
+        EntryT* entry = byte_it.index_ptr(group_entries);
+        if (LLVM_LIKELY(entry->key() == lookup_key)) {
+          __builtin_assume(entry != nullptr);
+          return entry;
+        }
+        ++byte_it;
+      } while (LLVM_UNLIKELY(byte_it != byte_end));
+    }
+
+    // We failed to find a matching entry in this bucket, so check if there are
+    // empty slots and we're done probing.
+    auto empty_byte_matched_range = g.MatchEmpty();
+    if (LLVM_LIKELY(empty_byte_matched_range)) {
+      return nullptr;
+    }
+
+    s.step();
+  } while (LLVM_UNLIKELY(true));
+}
+
+template <typename InputKeyT, typename InputValueT>
+template <typename EntryCallbackT, typename GroupCallbackT>
+[[clang::always_inline]] auto ViewImpl<InputKeyT, InputValueT>::ForEachEntry(
+    EntryCallbackT entry_callback, GroupCallbackT group_callback) const
+    -> void {
+  uint8_t* local_metadata = metadata();
+  EntryT* local_entries = entries();
+
+  ssize_t local_size = size();
+  for (ssize_t group_index = 0; group_index < local_size;
+       group_index += GroupSize) {
+    auto g = MetadataGroup::Load(local_metadata, group_index);
+    auto present_matched_range = g.MatchPresent();
+    if (!present_matched_range) {
+      continue;
+    }
+    for (ssize_t byte_index : present_matched_range) {
+      entry_callback(local_entries[group_index + byte_index]);
+    }
+
+    group_callback(&local_metadata[group_index]);
+  }
+}
+
+template <typename InputKeyT, typename InputValueT>
+auto ViewImpl<InputKeyT, InputValueT>::CountProbedKeys() const -> ssize_t {
```

Comment:
If you don't want to document what this is computing at the API level, you should document it here at its definition. It looks like this is counting the keys that are not in the first bucket of their probe sequence?

With this name, I could have imagined this counted number of probes in lookups in this hashmap, or the sum of how far each key is from the start of its probe sequence.